lib/edge-runtime/ai.ts

```diff
@@ -0,0 +1,580 @@
+import { Redis } from "@upstash/redis/nodejs";
+import {
+  ActionToHttpRequest,
+  AnswersType,
+  ChatGPTMessage,
+  ChatGPTParams,
+  GPTMessageInclSummary,
+  StreamingStepInput,
+  ToConfirm,
+} from "../models";
+import { ActionPlusApiInfo, OrgJoinIsPaidFinetunedModels } from "../types";
+import getMessages, { chatToDocsPrompt } from "../prompts/chatBot";
+import {
+  repopulateVariables,
+  sanitizeMessages,
+} from "./apiResponseSimplification";
+import { MessageInclSummaryToGPT, removeOldestFunctionCalls } from "./utils";
+import { exponentialRetryWrapper, getTokenCount, openAiCost } from "../utils";
+import { streamLLMResponse } from "../queryLLM";
+import { MAX_TOKENS_OUT } from "../consts";
+import { FunctionCall, parseOutput } from "@superflows/chat-ui-react";
+import { filterActions } from "./filterActions";
+import {
+  streamResponseToUser,
+  updatePastAssistantMessage,
+} from "./angelaUtils";
+import { getMissingArgCorrections } from "./missingParamCorrection";
+import { requestCorrectionSystemPrompt } from "../prompts/requestCorrection";
+import {
+  constructHttpRequest,
+  getDocsChatRequest,
+  makeHttpRequest,
+  processAPIoutput,
+} from "./requests";
+import summarizeText from "./summarize";
+import { createClient } from "@supabase/supabase-js";
+import { Database } from "../database.types";
+import {
+  getRelevantDocChunks,
+  getSearchDocsAction,
+} from "../embed-docs/docsSearch";
+
+let redis: Redis | null = null;
+if (process.env.UPSTASH_REDIS_REST_URL && process.env.UPSTASH_REDIS_REST_TOKEN)
+  redis = Redis.fromEnv();
+
+const supabase = createClient<Database>(
+  process.env.NEXT_PUBLIC_SUPABASE_URL!, // The existence of these is checked by answers
+  process.env.SERVICE_LEVEL_KEY_SUPABASE!,
+);
+
+export async function Dottie( // Dottie talks to docs
+  controller: ReadableStreamDefaultController,
+  reqData: AnswersType,
+  org: OrgJoinIsPaidFinetunedModels,
+  conversationId: number,
+  previousMessages: ChatGPTMessage[],
+  language: string | null,
+): Promise<{
+  nonSystemMessages: ChatGPTMessage[];
+  cost: number;
+  numUserQueries: number;
+}> {
+  console.log("Dottie called!");
+  const streamInfo = await preamble(
+    controller,
+    conversationId,
+    [],
+    reqData,
+    org,
+    language,
+  );
+
+  const model = org.model;
+  const nonSystemMessages = [...previousMessages];
+  const maxConvLength = model === "gpt-4-0613" ? 20 : 14;
+  let cost = 0;
+
+  // Const controlling how much context to include
+  const nChunksInclude = 3;
+
```

Comment:
so, with the removal of the "more documentation" logic, what do we do if the answer is in the 4th most relevant chunk?